src/preprocessing/passes/non_clausal_simp.cpp

```diff
@@ -448,10 +432,105 @@ PreprocessingPassResult NonClausalSimp::applyInternal(
         Rewriter::rewrite(Node(learnedBuilder)));
   }
 
-  propagator->setNeedsFinish(true);
   return PreprocessingPassResult::NO_CONFLICT;
 }  // namespace passes
 
+std::pair<bool, std::vector<Node>> NonClausalSimp::preprocessByCryptoMinisat(
+    AssertionPipeline* assertionsToPreprocess, unsigned substs_index)
+{
+  std::unique_ptr<SatSolver> satSolver(SatSolverFactory::createCryptoMinisat(
+      smtStatisticsRegistry(), "non_clausal_simp_solver"));
+  CVC4::prop::NullRegistrar registrar;
+  context::Context context;
+  CVC4::prop::TseitinCnfStream cnfStream(satSolver.get(),
+                                         &registrar,
+                                         &context,
+                                         true,
+                                         "toCNF-skeleton-preprocessing");
+  std::vector<Node> learned_literals;
+  Trace("non-clausal-simplify") << "asserting to sat solver" << std::endl;
+  {  // for timing
+    TimerStat::CodeTimer timer(d_statistics.d_cnfTranslateTime);
+    for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+    {
+      Assert(Rewriter::rewrite((*assertionsToPreprocess)[i])
+             == (*assertionsToPreprocess)[i]);
+      // Don't reprocess substitutions
+      if (substs_index > 0 && i == substs_index)
+      {
+        continue;
+      }
+      Trace("non-clausal-simplify")
+          << "asserting " << (*assertionsToPreprocess)[i] << std::endl;
+      cnfStream.convertAndAssert(
+          (*assertionsToPreprocess)[i], false, false, RULE_GIVEN);
+    }
+  } // end timing
+  Trace("non-clausal-simplify") << "solving" << std::endl;
+  if (satSolver->solve() == SAT_VALUE_FALSE)
+  {
+    // If in conflict, just return false
+    return std::make_pair(false, learned_literals);
+  }
+
+  std::vector<SatLiteral> topLevelUnits = satSolver->getTopLevelUnits();
+  for (size_t i = 0; i < topLevelUnits.size(); i++)
+  {
+    SatLiteral lit = topLevelUnits[i];
+    if (const TNode* learnedLiteral = FindOrNull(cnfStream.getNodeCache(), lit))
+    {
+      // If the learned literal is not created internally by cryptominisat
+      learned_literals.push_back(*learnedLiteral);
+    }
+  }
+  return std::make_pair(true, learned_literals);
+}
+
+std::pair<bool, std::vector<Node>>
+NonClausalSimp::preprocessByCircuitPropagator(
+    AssertionPipeline* assertionsToPreprocess, unsigned substs_index)
+{
+  theory::booleans::CircuitPropagator* propagator =
+      d_preprocContext->getCircuitPropagator();
+  std::vector<Node> learned_literals;
+  if (propagator->getNeedsFinish())
+  {
+    propagator->finish();
+    propagator->setNeedsFinish(false);
+  }
+  propagator->initialize();
+
+  // Assert all the assertions to the propagator
+  Trace("non-clausal-simplify") << "asserting to propagator" << std::endl;
+  {  // for timing
+    TimerStat::CodeTimer timer(d_statistics.d_cnfTranslateTime);
```

Comment:
If you also use `d_statistics.d_cnfTranslateTime` for the circuit propagator the name of the statistic is not appropriate since the circuit propagator does not do any CNF translation. In both cases you add assertions maybe call it `addAssertionTime` then?